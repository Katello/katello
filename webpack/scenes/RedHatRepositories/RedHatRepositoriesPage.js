/* eslint-disable import/no-extraneous-dependencies */
/* eslint import/no-unresolved: [2, { ignore: [foremanReact/*] }] */
/* eslint-disable import/no-unresolved */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';
import { Grid, Row, Col } from 'react-bootstrap';
import { Skeleton, Alert } from '@patternfly/react-core';
import { Button } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';
import PermissionDenied from 'foremanReact/components/PermissionDenied';
import { LoadingState } from '../../components/LoadingState';
import { createEnabledRepoParams } from '../../redux/actions/RedHatRepositories/enabled';
import SearchBar from './components/SearchBar';
import RecommendedRepositorySetsToggler from './components/RecommendedRepositorySetsToggler';
import { getSetsComponent, getEnabledComponent } from './helpers';
import api from '../../services/api';
import { AIRGAPPED } from '../Subscriptions/Manifest/CdnConfigurationTab/CdnConfigurationConstants';

class RedHatRepositoriesPage extends Component {
  componentDidMount() {
    this.loadData();
  }

  loadData() {
    this.props.loadOrganization();
    this.props.loadEnabledRepos();
    this.props.loadRepositorySets({ search: { filters: ['rpm'] } });
  }

  render() {
    const { enabledRepositories, repositorySets, organization } = this.props;
    const { repoParams } = createEnabledRepoParams(enabledRepositories);

    if (!isEmpty(repositorySets.missingPermissions)) {
      return <PermissionDenied missingPermissions={repositorySets.missingPermissions} />;
    }
    if (!isEmpty(enabledRepositories.missingPermissions)) {
      return <PermissionDenied missingPermissions={enabledRepositories.missingPermissions} />;
    }
    if (!(organization?.cdn_configuration)) {
      return <Skeleton />;
    }
    if (organization.cdn_configuration.type === AIRGAPPED) {
      return (
        <Grid id="redhatRepositoriesPage" bsClass="container-fluid">
          <h1>{__('Red Hat Repositories')}</h1>
          <Row className="toolbar-pf">
            <Col>
              <Alert
                variant="info"
                className="repo-sets-alert"
                isInline
                title={__('CDN configuration is set to Air-gapped (disconnected). Repository enablement/disablement is not permitted on this page.')}
              />
            </Col>
          </Row>
        </Grid>
      );
    }

    return (
      <Grid id="redhatRepositoriesPage" bsClass="container-fluid">
        <h1>{__('Red Hat Repositories')}</h1>
        <Row className="toolbar-pf">
          <Col sm={12}>
            <SearchBar />
          </Col>
        </Row>

        <Row className="row-eq-height">
          <Col sm={6} className="available-repositories-container">
            <div className="available-repositories-header">
              <h2>{__('Available Repositories')}</h2>
              <RecommendedRepositorySetsToggler
                enabled={repositorySets.recommended}
                onChange={value => this.props.updateRecommendedRepositorySets(value)}
                className="recommended-repositories-toggler"
              />
            </div>
            <LoadingState loading={repositorySets.loading} loadingText={__('Loading')}>
              {getSetsComponent(
                repositorySets,
                (pagination) => {
                  this.props.loadRepositorySets({
                    ...pagination,
                    search: repositorySets.search,
                  });
                },
              )}
            </LoadingState>
          </Col>

          <Col sm={6} className="enabled-repositories-container">
            <h2>
              {__('Enabled Repositories')}
              <Button
                className="pull-right"
                onClick={() => { api.open('/repositories.csv', repoParams); }}
              >
                {__('Export as CSV')}
              </Button>
            </h2>

            <LoadingState loading={enabledRepositories.loading} loadingText={__('Loading')}>
              {getEnabledComponent(
                enabledRepositories,
                (pagination) => {
                  this.props.loadEnabledRepos({
                    ...pagination,
                    search: enabledRepositories.search,
                  });
                },
              )}
            </LoadingState>
          </Col>
        </Row>
      </Grid>
    );
  }
}

RedHatRepositoriesPage.propTypes = {
  loadOrganization: PropTypes.func.isRequired,
  loadEnabledRepos: PropTypes.func.isRequired,
  loadRepositorySets: PropTypes.func.isRequired,
  updateRecommendedRepositorySets: PropTypes.func.isRequired,
  enabledRepositories: PropTypes.shape({
    loading: PropTypes.bool,
    search: PropTypes.shape({}),
    missingPermissions: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  repositorySets: PropTypes.shape({
    recommended: PropTypes.bool,
    loading: PropTypes.bool,
    search: PropTypes.shape({}),
    // Disabling rule as existing code failed due to an eslint-plugin-react update
    // eslint-disable-next-line react/forbid-prop-types
    missingPermissions: PropTypes.array,
  }).isRequired,
  organization: PropTypes.shape({
    cdn_configuration: PropTypes.shape({
      type: PropTypes.string,
    }),
  }).isRequired,

};

export default RedHatRepositoriesPage;
