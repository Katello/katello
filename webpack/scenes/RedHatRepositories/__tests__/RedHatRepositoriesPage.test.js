import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RedHatRepositoriesPage from '../RedHatRepositoriesPage';

jest.mock('foremanReact/components/PermissionDenied');

describe('RedHatRepositories page', () => {
  const page = shallow(<RedHatRepositoriesPage
    loadEnabledRepos={jest.fn()}
    loadRepositorySets={jest.fn()}
    loadOrganization={jest.fn()}
    updateRecommendedRepositorySets={jest.fn()}
    enabledRepositories={{
      loading: false,
      search: {},
      missingPermissions: [],
      repositories: [],
    }}
    repositorySets={{
      recommended: false,
      loading: false,
      search: {},
      missingPermissions: [],
    }}
    organization={{
      id: 1000,
      cdn_configuration: {
        type: 'redhat_cdn',
        url: 'http://cdn.example.com',
      },
    }}
  />);

  const permissionDeniedPage = shallow(<RedHatRepositoriesPage
    loadEnabledRepos={jest.fn()}
    loadRepositorySets={jest.fn()}
    loadOrganization={jest.fn()}
    updateRecommendedRepositorySets={jest.fn()}
    enabledRepositories={{
      loading: false,
      search: {},
      missingPermissions: ['view_organizations'],
    }}
    repositorySets={{
      recommended: false,
      loading: false,
      search: {},
      missingPermissions: ['view_organizations'],
    }}
    organization={{
      id: 1000,
      cdn_configuration: {
        type: 'redhat_cdn',
        url: 'http://cdn.example.com',
      },
    }}
  />);

  it('should render', async () => {
    expect(toJson(page)).toMatchSnapshot();
  });

  it('should render <PermissionDenied /> when permissions are missing', async () => {
    expect(toJson(permissionDeniedPage)).toMatchSnapshot();
  });
});
