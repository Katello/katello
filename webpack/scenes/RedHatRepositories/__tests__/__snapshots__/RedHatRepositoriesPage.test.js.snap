// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedHatRepositories page should render <PermissionDenied /> when permissions are missing 1`] = `
<PermissionDenied
  missingPermissions={
    Array [
      "view_organizations",
    ]
  }
  primaryButton={null}
/>
`;

exports[`RedHatRepositories page should render 1`] = `
<Grid
  bsClass="container-fluid"
  componentClass="div"
  fluid={false}
  id="redhatRepositoriesPage"
>
  <h1>
    Red Hat Repositories
  </h1>
  <Row
    bsClass="row"
    className="toolbar-pf"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={12}
    >
      <Connect(SearchBar) />
    </Col>
  </Row>
  <Row
    bsClass="row"
    className="row-eq-height"
    componentClass="div"
  >
    <Col
      bsClass="col"
      className="available-repositories-container"
      componentClass="div"
      sm={6}
    >
      <div
        className="available-repositories-header"
      >
        <h2>
          Available Repositories
        </h2>
        <RecommendedRepositorySetsToggler
          className="recommended-repositories-toggler"
          enabled={false}
          help="This shows repositories that are used in a typical setup."
          onChange={[Function]}
        >
          Recommended Repositories
        </RecommendedRepositorySetsToggler>
      </div>
      <LoadingState
        loading={false}
        loadingText="Loading"
        timeout={300}
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "No Red Hat products currently exist, please import a manifest <a href=\\"/subscriptions/\\"> here </a> to receive Red Hat content. No repository sets available.",
            }
          }
        />
      </LoadingState>
    </Col>
    <Col
      bsClass="col"
      className="enabled-repositories-container"
      componentClass="div"
      sm={6}
    >
      <h2>
        Enabled Repositories
        <FieldLevelHelp
          buttonClass=""
          content="Only repositories not published in a content view can be disabled. Published repositories must be deleted from the repository details page."
          placement="top"
          rootClose={true}
        />
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="pull-right"
          disabled={false}
          onClick={[Function]}
          ouiaId="export-csv-button"
        >
          Export as CSV
        </Button>
      </h2>
      <LoadingState
        loading={false}
        loadingText="Loading"
        timeout={300}
      >
        <ListView
          className=""
        >
          <div
            className="sticky-pagination sticky-pagination-grey"
          >
            <Pagination
              className={null}
              isCompact={true}
              itemCount={0}
              noSidePadding={false}
              onChange={[Function]}
              onPerPageSelect={null}
              onSetPage={null}
              page={1}
              perPage={null}
              updateParamsByUrl={false}
              variant="bottom"
            />
          </div>
        </ListView>
      </LoadingState>
    </Col>
  </Row>
</Grid>
`;
